demo_nodes_cpp/src/parameters/parameter_events.cpp

```diff
@@ -82,7 +88,10 @@ int main(int argc, char ** argv)
 
   // TODO(wjwwood): Create and use delete_parameter
 
-  rclcpp::sleep_for(100ms);
+  // TODO(hidmic): Fast-RTPS takes a significant amount of time to deliver
+  //               requests and response, thus the rather long sleep. Reduce
+  //               once that's resolved.
+  rclcpp::sleep_for(3s);
```

Comment:
Why is this sleep needed at all? We should use a future and `spin_until_future_complete()` if this is now causing us flakyness.

Reply:
That is what's happening, see above we're using a `SyncParametersClient`. However, without a sleep of at least 3 seconds, the request would never arrive nor be served during the `spin_some()` immediately below whenever using `rmw_fastrtps_cpp` as `RMW_IMPLEMENTATION`.

This is not a fix but a patch, to not block these PRs on unrelated failures that IIUC are not subject to the API freeze.

Reply:
Similar to the other comment, what worries me about this is that we cannot explain why this change is needed with the new system (I assume it is not failing on master's CI). I think we should either figure out why the launch changes require this increase and avoid that reason, or we should clear up the technical debt (not make it worse).

Reply:
I can't tell either why CI is not failing on `master` at the moment, but I've run the binaries manually on a terminal and I've seen them misbehave. Once I get all these PRs to pass CI, likely tomorrow, I'll dig into the old code to figure out why it was succeeding. It wouldn't be the first time we have tests that are not testing anything (e.g. [this one](https://github.com/ros2/launch/pull/167#issuecomment-465586945)).

Reply:
Well you don't have to spend a lot of time figuring out why, but what I meant is that if the behavior of this test changed, it might be a red flag that some change in the testing system is having an adverse affect on the tests. My reaction to that would be "why is this test now breaking" not "what do I change in this test to get it to pass again". That's my concern here.

If it ends up being something we cannot explain (not ideal) or basically "timing" then we can patch this test, but it's not ideal. I'd also be ok with "doesn't matter why this is _now_ failing, as the test had technical debt from before (the sleep), so we'll fix it up without investigating the root cause of this becoming more flaky".

Reply:
> My reaction to that would be "why is this test now breaking" not "what do I change in this test to get it to pass again".

That was my reaction as well. Which led me to reproduce the test (and the failure) manually. And by manually I mean running the binaries on a terminal, with no launch automation at all. Sorry for not conveying that piece of information properly.

> If it ends up being something we cannot explain (not ideal) or basically "timing" then we can patch this test, but it's not ideal

I know it's not ideal and it was not my intention to propose this as a fix. It's just a short term patch for issues unrelated to the new launch testing code (or that the new launch testing code deals poorly with, see [here](https://github.com/ros2/demos/pull/318/files#diff-af9f8c2a880ea8ef2e60c0ac41355245R55) for an example), so that we can get this in by Thursday. What comes immediately after is a follow-up and investigation on each of the failures. Some I managed to narrow down already (https://github.com/ros2/rclpy/issues/327, for instance), but others are harder to debug, specially the ones that have to do with timing (most if not all the ones I just patched).

Reply:
I've reproduced the failure on `master` i.e. when using Fast-RTPS, parameter event demo nodes terminate early with no output. As to why master is passing CI anyways, I've just found a bug in the current `launch_testing` code: there should be a `return` statement right after calling `self._fail` [here]( https://github.com/ros2/launch/blob/master/launch_testing/launch_testing/__init__.py#L211-L217).

Reply:
> that the new launch testing code deals poorly with, see [here](https://github.com/ros2/demos/pull/318/files#diff-af9f8c2a880ea8ef2e60c0ac41355245R55) for an example

I can reproduce this one on `master` too. If using a somewhat bulkier RMW implementation like Connext, client demo nodes exit with non zero return codes if Ctrl-C is sent right after the server response, likely during shutdown. Which is what the new launch testing code is attempting to do without that delay.